File: leetcode/0264-ugly-number-ii/solution.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        ugly = [1]
        p2 = p3 = p5 = 0

        while len(ugly) < n:
            next2, next3, next5 = ugly[p2] * 2, ugly[p3] * 3, ugly[p5] * 5
            next_ugly = min(next2, next3, next5)
            ugly.append(next_ugly)

            if next_ugly == next2:
                p2 += 1
            if next_ugly == next3:
                p3 += 1
            if next_ugly == next5:
                p5 += 1

        return ugly[-1]

File: leetcode/0264-ugly-number-ii/solution.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int nthUglyNumber(int n) {
        vector<long long> ugly(n);
        ugly[0] = 1;
        int p2 = 0, p3 = 0, p5 = 0;

        for (int i = 1; i < n; i++) {
            long long next2 = ugly[p2] * 2;
            long long next3 = ugly[p3] * 3;
            long long next5 = ugly[p5] * 5;
            long long nextUgly = min({next2, next3, next5});
            ugly[i] = nextUgly;

            if (nextUgly == next2) p2++;
            if (nextUgly == next3) p3++;
            if (nextUgly == next5) p5++;
        }

        return (int)ugly[n - 1];
    }
};
